reactor-core/src/main/java/reactor/core/publisher/FluxConcatMapNoPrefetch.java

```diff
@@ -0,0 +1,299 @@
+		}
+
+		@Override
+		public CoreSubscriber<? super R> actual() {
+			return actual;
+		}
+
+		@Override
+		public void onSubscribe(Subscription s) {
+			if (Operators.validate(this.upstream, s)) {
+				this.upstream = s;
+
+				actual.onSubscribe(this);
+			}
+		}
+
+		@Override
+		public void onNext(T t) {
+			if (!STATE.compareAndSet(this, State.REQUESTED, State.ACTIVE)) {
+				switch (state) {
+					case CANCELED:
+						Operators.onDiscard(t, currentContext());
+						break;
+					case TERMINATED:
+						Operators.onNextDropped(t, currentContext());
+						break;
+				}
+				return;
+			}
+
+			try {
+				Publisher<? extends R> p = mapper.apply(t);
+				Objects.requireNonNull(p, "The mapper returned a null Publisher");
+
+				if (p instanceof Callable) {
+					Callable<R> callable = (Callable<R>) p;
+
+					R result = callable.call();
+					if (result == null) {
+						innerComplete();
+						return;
+					}
+
+					if (inner.isUnbounded()) {
+						actual.onNext(result);
+						innerComplete();
+						return;
+					}
+
+					inner.set(new WeakScalarSubscription<>(result, inner));
+					return;
+				}
+
+				p.subscribe(inner);
+			}
+			catch (Throwable e) {
+				Context ctx = actual.currentContext();
+				Operators.onDiscard(t, ctx);
+
+				Throwable e_ = Operators.onNextError(t, e, ctx, upstream);
+				if (e_ == null) {
+					innerComplete();
+					return;
+				}
+				STATE.lazySet(this, State.TERMINATED);
+				actual.onError(Operators.onOperatorError(upstream, e, t, ctx));
+				return;
+			}
+		}
+
+		@Override
+		public void onError(Throwable t) {
+			for (State previousState = this.state; ; previousState = this.state) {
+				switch (previousState) {
+					case CANCELED:
+					case TERMINATED:
+						Operators.onErrorDropped(t, currentContext());
+						return;
+					default:
+						if (!STATE.compareAndSet(this, previousState, State.TERMINATED)) {
+							continue;
+						}
+						inner.cancel();
+						synchronized (this) {
+							actual.onError(t);
+						}
+						return;
+				}
+			}
+		}
+
+		@Override
+		public void onComplete() {
+			for (State previousState = this.state; ; previousState = this.state) {
+				switch (previousState) {
+					case INITIAL:
+					case REQUESTED:
+						if (!STATE.compareAndSet(this, previousState, State.TERMINATED)) {
+							continue;
+						}
+						actual.onComplete();
+						return;
+					case ACTIVE:
+						if (!STATE.compareAndSet(this, previousState, State.LAST_ACTIVE)) {
+							continue;
+						}
+						return;
+					default:
+						return;
+				}
+			}
+		}
+
+		@Override
+		public synchronized void innerNext(R value) {
+			switch (state) {
+				case ACTIVE:
+				case LAST_ACTIVE:
+					actual.onNext(value);
+					break;
+				default:
+					Operators.onDiscard(value, currentContext());
+					break;
+			}
+		}
+
+		@Override
+		public void innerComplete() {
+			for (State previousState = this.state; ; previousState = this.state) {
+				switch (previousState) {
```

Comment:
(this generally applies to CAS in that class)

Reply:
I thought about doing it, but then realized that it is not copy/paste friendly and we may slip some bugs due to it.
When reviewing, having to check `previousState` IMO is actually better, because it forces you to double check that we're in the correct `case` branch. Otherwise imagine accidentally approving something like:
```java
switch (previousState) {
	case ACTIVE:
		if (!STATE.compareAndSet(this, ACTIVE, State.REQUESTED)) {
			continue;
		}
		upstream.request(1);
		return;
	case LAST_ACTIVE:
		if (!STATE.compareAndSet(this, ACTIVE, State.TERMINATED)) {
			continue;
		}
		actual.onComplete();
		return;
	default:
		return;
}
```

if not covered with tests (hehe :D), we will ship a potentially infinite loop to production, and the cost of such mistake would be high

Reply:
ok, that makes sense as well. let's keep it that way then.